### ISSUE TITLE
currentPage: 0, currentPage: 1 and currentPage: -1 produces the same output for products query when filtering is used

### ISSUE THREAD

**Issue by TomashKhamlai** [magento/graphql-ce#727]
### Preconditions (*)

1. Import products
[import.csv.zip](https://github.com/magento/graphql-ce/files/3231912/import.csv.zip)

If this doesn't work then modify the line 8 in the script to replace my URL to yours. Make it executable and run it.
[test_pager_graphql.sh.zip](https://github.com/magento/graphql-ce/files/3231820/test_pager_graphql.sh.zip)

### Steps to reproduce (*)

1. Compare the output of these 3 queries:
```graphql
{
  products(filter: {sku: {notnull:""}}, currentPage: 0, pageSize: 20) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      sku
    }
  }
}
```
```graphql
{
  products(filter: {sku: {notnull:""}}, currentPage: 1, pageSize: 20) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      sku
    }
  }
}
```
```graphql
{
  products(filter: {sku: {notnull:""}}, currentPage: -1, pageSize: 20) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      sku
    }
  }
}
```

### Expected result (*)

1. When the current page is 0 the output should contain the message that the page number is not valid

### Actual result (*)

The output is almost the same only page number is different
```graphql
{
  "data": {
    "products": {
      "page_info": {
        "current_page": -1,
        "page_size": 20,
        "total_pages": 4
      },
      "total_count": 67,
      "items": [
        {
          "sku": "jacket-abbibass-1"
        },
        {
          "sku": "jacket-abbibass-2"
        },
        {
          "sku": "jacket-abbibass-3"
        },
        {
          "sku": "elvis-t-shirt-(sale)-1"
        },
        {
          "sku": "elvis-t-shirt-(sale)-2"
        },
        {
          "sku": "elvis-t-shirt-(sale)-3"
        },
        {
          "sku": "elvis-t-shirt-(sale)-4"
        },
        {
          "sku": "elvis-t-shirt-(sale)-5"
        },
        {
          "sku": "elvis-t-shirt-(sale)-6"
        },
        {
          "sku": "elvis-t-shirt-(sale)-7"
        },
        {
          "sku": "elvis-t-shirt-(sale)-8"
        },
        {
          "sku": "elvis-t-shirt-(sale)-9"
        },
        {
          "sku": "elvis-t-shirt-(sale)-10"
        },
        {
          "sku": "elvis-t-shirt-(sale)-11"
        },
        {
          "sku": "elvis-t-shirt-(sale)-12"
        },
        {
          "sku": "elvis-t-shirt-(sale)-13"
        },
        {
          "sku": "elvis-t-shirt-(sale)-14"
        },
        {
          "sku": "elvis-t-shirt-(sale)-15"
        },
        {
          "sku": "elvis-t-shirt-(sale)-16"
        },
        {
          "sku": "elvis-t-shirt-(sale)-17"
        }
      ]
    }
  }
}
```

**Comment by cjelger** [comment]
@TomashKhamlai From what I have seen and the discussion we had on Slack, the problem that I witnessed is not just that the page number can be less than 1. The problem is that the `filter` search properly starts its pagination at page 1, but the products `search` starts its pagination at page 0.

For example, with the Luma sample data and the following query:
```
{
  products(search: "yoga", currentPage: 0, pageSize: 20) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      sku
    }
  }
}
```
I get very weird results with Magento 2.3.1. I get for example a `total_count` of 35, page 0 has the first 20 results (yes page 0, not page 1!), page 1 also has 20 results, and page 2 has the same 20 results than page 1 ... while the total count is 35 ...

**Comment by cjelger** [comment]
I spent some more time on that, also comparing the results I get in the admin backend. So again with Magento 2.3.1 and the Luma sample data, I searched the "yoga" products in the admin products panel and filtered them with "visibility: catalog, search", and sorted them by ascending SKU. I get 35 results.

Now if I compare this outcome with the following query result:
```
{
  products(search: "yoga", currentPage: 1, pageSize: 20, sort: {sku: ASC}) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      sku
    }
  }
}
```
I get the same 20 products than in the admin panel, and a `total_count` of 35. So far so good. Let's assume we consider these products as the products with IDs 1 to 20.

Now the same GraphQL query with `currentPage: 2` returns the same products as for page 1 BUT shifted by a single product. In terms of IDs, this means we get products with IDs 2 to 21.

Now the same GraphQL query with `currentPage: 0` returns the products with IDs 16 to 35, so the last 5 products of page 1 and the 15 products that should appear on page 2.

There is definitely something wrong with the pagination, but there is some kind of a pattern in the way the results are wrongly returned, so I suspect a silly calculation issue for the pagination.

**Comment by TomashKhamlai** [comment]
@cjelger, thank you for the information. I will check this white testing PR.

**Comment by lenaorobei** [comment]
@cjelger I cannot reproduce this on latest `2.3-develop`.

Query
``` js
{
  products(search: "yoga", currentPage: 1, pageSize: 5, sort: {sku: ASC}) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      id
      sku
    }
  }
}
```
Response
``` js
  "data": {
    "products": {
      "page_info": {
        "current_page": 1,
        "page_size": 5,
        "total_pages": 7
      },
      "total_count": 35,
      "items": [
        {
          "id": 29,
          "sku": "24-MB03"
        },
        {
          "id": 33,
          "sku": "24-UB02"
        },
        {
          "id": 34,
          "sku": "24-WB01"
        },
        {
          "id": 35,
          "sku": "24-WB02"
        },
        {
          "id": 38,
          "sku": "24-WB03"
        }
      ]
    }
  }
}
```
Query 
``` js
{
  products(search: "yoga", currentPage: 2, pageSize: 5, sort: {sku: ASC}) {
    page_info {
      current_page
      page_size
      total_pages
    }
    total_count
    items {
      id
      sku
    }
  }
}
```
Response
``` js
{
  "data": {
    "products": {
      "page_info": {
        "current_page": 2,
        "page_size": 5,
        "total_pages": 7
      },
      "total_count": 35,
      "items": [
        {
          "id": 78,
          "sku": "24-WG080"
        },
        {
          "id": 47,
          "sku": "24-WG084"
        },
        {
          "id": 71,
          "sku": "24-WG085_Group"
        },
        {
          "id": 48,
          "sku": "24-WG088"
        },
        {
          "id": 72,
          "sku": "240-LV04"
        }
      ]
    }
  }
}
```

**Comment by cjelger** [comment]
@lenaorobei Yes the problem always appear on the last page, so in your case, with `currentPage: 7` you get the wrong results. In the mean time I found the problem. This line here is wrong:

https://github.com/magento/graphql-ce/blob/5d4520baea3712ef44e1f54c4291e8ce3b193f2e/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php#L147

it should be without the `-1` at the end:

`$maxPages = ceil($searchResult->getTotalCount() / $searchCriteria->getPageSize());`

@pmclain I stopped developing with PHP when PHP4 was out, and I'm not a Magento developer, do you mind fixing this in your open PR? That `-1` is wrong here because this method uses the GraphQL pagination starting at page 1, and not a pagination starting at 0. This fixes the issue, at least with the Luma example, with page size 20 or 5 as mentioned by @lenaorobei .

**Comment by cjelger** [comment]
@naydav Do I understand correctly that this fix will only be available in 2.3.3 or will it also go in 2.3.2?
